sqlmesh/core/engine_adapter/snowflake.py

```diff
@@ -326,7 +345,39 @@ def _get_data_objects(
         ]
 
     def set_current_catalog(self, catalog: str) -> None:
-        self.execute(exp.Use(this=exp.to_identifier(catalog)))
+        self.execute(exp.Use(this=normalize_identifiers(catalog, dialect=self.dialect)))
+
+    def set_current_schema(self, schema: str) -> None:
+        self.execute(exp.Use(kind="SCHEMA", this=to_schema(schema)))
+
+    def _to_sql(self, expression: exp.Expression, quote: bool = True, **kwargs: t.Any) -> str:
+        # note: important to use self._default_catalog instead of the self.default_catalog property
+        # otherwise we get RecursionError: maximum recursion depth exceeded
+        # because it calls get_current_catalog(), which executes a query, which needs the default catalog, which calls get_current_catalog()... etc
+        if self._default_catalog:
+
+            def unquote_and_lower(identifier: str) -> str:
+                return exp.to_identifier(identifier).name.lower()
+
+            default_catalog_unquoted = unquote_and_lower(self._default_catalog)
+            default_catalog_normalized = normalize_identifiers(
+                self._default_catalog, dialect=self.dialect
+            )
+
+            def catalog_rewriter(node: exp.Expression) -> exp.Expression:
+                if isinstance(node, exp.Table):
+                    if node.catalog:
+                        # only replace the catalog on the model with the target catalog if the two are functionally equivalent
+                        if unquote_and_lower(node.catalog) == default_catalog_unquoted:
+                            node.set("catalog", default_catalog_normalized)
+                return node
+
+            # Rewrite whatever default catalog is present on the query to be compatible with what the user supplied in the
```

Comment:
I thought that was by design? And that's exactly why we required users to specify the catalog in quotes in the connection config when using a cross-dialect setup. @eakmanrq please correct me if I'm wrong.

Reply:
So what happens if you specify `database: '"ERIN"'` in the Snowflake connection config is [this code](https://github.com/TobikoData/sqlmesh/blob/7acaef31817a4481666e7a109daa0c1c019f61b8/sqlmesh/core/model/meta.py#L471) completely ignores what you did, strips the quotes and re-normalizes it in the source dialect. The normalized identifier in the source dialect is not compatible with Snowflake.

The reason why we dont just update `model.fqn` to fix the catalog is because `model.fqn` is used as a key all over the place in internal dictionaries and potentially the state

Reply:
I would like to hear more about what the ideal fix would be for `model.fqn`. Are you suggesting mixing the reference dialect used for normalization so that catalog is from target and db/table is from source? Or it all uses target instead?

Reply:
Yes that's right. So the model `fqn` is currently `<catalog>.<schema>.<table>`, all normalized and quoted in the source dialect.

The proposal would be to leabe `<schema>.<table>` normalized and quoted in the source dialect, but change `<catalog>` to be normalized and quoted in the _target_ dialect, but ONLY if `<catalog>` came from the gateway config. If `<catalog>` was manually specified on the model by the user, then we of don't touch it

Reply:
Ok so after talking to @izeigerman, and going down a massive rabbit hole trying to normalize the `<catalog>` part to the target dialect and the rest to the source dialect, we determined it was easier to just do the catalog replacement within the adapter (like in this PR) because the change is minimal and also self contained to the adapter.

If anyone's interested, fixing `model.fqn` to normalize the `<catalog>` part  separately from the rest is harder than it sounds. Even if you fix all the tests that make assumptions about the format of the catalog you're still left with broken dependencies.

This is because `model.depends_on` [runs the model query and extracts the table names](https://github.com/TobikoData/sqlmesh/blob/e27daf54c19ba1c3d0fe78a86f5b928b3795630e/sqlmesh/core/model/definition.py#L846) (which [of course](https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/renderer.py#L236) are [fully normalized in the source dialect](https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/renderer.py#L295), [including the catalog](https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/dialect.py#L1079)). Due to this, it then fails to match them against the models loaded into the Context (because `model.fqn` doesnt match against what was extracted from the query), so the dependency tree is broken.